pilot/pkg/controllers/ipallocate/ipallocate.go

```diff
@@ -0,0 +1,402 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipallocate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/netip"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch/v5"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/api/meta/v1alpha1"
+	networkingv1alpha3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	autoallocate "istio.io/istio/pilot/pkg/serviceregistry/serviceentry"
+	kubelib "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/slices"
+)
+
+var log = istiolog.RegisterScope("ip-autoallocate", "IP autoallocate controller")
+
+type IPAllocate struct {
+	serviceEntryClient kclient.Client[*networkingv1alpha3.ServiceEntry]
+	stopChan           <-chan struct{}
+	queue              controllers.Queue
+
+	// This controller is not safe for concurency but it exists outside the critical path performing important but minimal functionality in a single thread.
+	// If we want the multi-thread this controller we must add locking around accessing the allocators which would otherwise be racy
+	v4allocator *ipAllocator
+	v6allocator *ipAllocator
+}
+
+type conflictDetectedEvent struct {
+	conflictingResourceIdentifier types.NamespacedName
+	conflictingAddress            netip.Addr
+}
+
+const (
+	controllerName = "IP Autoallocate"
+	// these are the ranges of the v1 logic, do we need to choose new ranges?
+	IPV4Prefix = "240.240.0.0/16"
```

Comment:
We ABSOLUTELY must change the ranges - and add a flag for admin to specify the range. 

We don't own the 240.240.0.0 range - IETF didn't assign it to Istio and it's not based on 'squatter rights' -the net admins may use this range for other purpose. 

Also since we do per-cluster allocs - the admin may want to manage the space and use different ranges for each cluster, in particular if combined with external DNS integrations. 

Same for the IPv6 prefix. 

If we add the flag - it would be ideal to allow users to specify one or the other - if they only specify an IPv6 CIDR we should only allocate IPv6, and the other way. Having dual stack or not is not an indication the user wants to allocate both IP4 and 6.

Reply:
To what exactly?

Reply:
I mean as the default, adding options is fine

Reply:
My preference: don't set any default for IPv4, set a FC:00/7 range for IPv6, 
and change the allocator logic for IPv6 to use truncated SHA of hostname or random ( there are 121 bits available, very low collision chance ). I didn't comment on the code - but I don't think for IPv6 64 or 121 bits we should increment a counter ( still need collision detection to be extra safe).

IPv4 space is VERY constrained - 240.240.0.0/16 is likely in use or will be in use, and it would be quite risky to claim any range that is in use. If admin wants auto-allocation - they should tell us what range they don't use. 

Or we can ask IETF to give us a private IPv4 range :-)

Reply:
To be clear - in the demo profile or in kind we can use 240.240.0.0 or even 10.0.0/8 - it's for demo. 

But I can't imagine a production user where Istio claiming a range that may be in use in the network is acceptable and safe. Yes, it's going to be hard for user to find an unused area - but that's how IPv4 needs to be handled those days.